core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -834,81 +832,6 @@ class KafkaApis(val requestChannel: RequestChannel,
       }
     }
 
-    def maybeConvertFetchedData(tp: TopicIdPartition,
-                                partitionData: FetchResponseData.PartitionData): FetchResponseData.PartitionData = {
-      // We will never return a logConfig when the topic is unresolved and the name is null. This is ok since we won't have any records to convert.
-      val logConfig = replicaManager.getLogConfig(tp.topicPartition)
-
-      if (logConfig.exists(_.compressionType == BrokerCompressionType.ZSTD) && versionId < 10) {
-        trace(s"Fetching messages is disabled for ZStandard compressed partition $tp. Sending unsupported version response to $clientId.")
-        FetchResponse.partitionResponse(tp, Errors.UNSUPPORTED_COMPRESSION_TYPE)
```

Comment:
I removed this check because it's kind of pointless (this topic config is rarely used and we never checked for the most common case of compression happening on the producer) and added unnecessarily complexity once down conversion was removed. Let me know if this is reasonable.

Reply:
> IIUC, we aren't doing on-the-fly down conversion anymore since we've dropped all of Fetch RPC versions that didn't support record version 2.

Correct.

> And for extant data with the older record versions, they are forward compatible with any newer Fetch RPC.

Correct, KIP-724 explains this in more detail. Unfortunately, the consumers need to support all record versions for now. See "Future Work" for more on what we would have to do to change that.